File: src/features/system-health/DiskMonitor.ts
import { Shell } from "../../utility/Shell.js";
import { Logger } from "../../utility/Logger.js";

/**
 * Disk usage information
 */
export interface DiskUsage {
	device: string;
	mountPoint: string;
	filesystem: string;
	totalBytes: number;
	usedBytes: number;
	freeBytes: number;
	percentUsed: number;
	totalGB: number;
	usedGB: number;
	freeGB: number;
	status: "healthy" | "warning" | "critical";
}

/**
 * Disk health warning
 */
export interface DiskWarning {
	device: string;
	mountPoint: string;
	level: "warning" | "critical";
	message: string;
	freeGB: number;
	percentUsed: number;
}

/**
 * SMART health status
 */
export interface SmartStatus {
	device: string;
	passed: boolean;
	temperature?: number;
	powerOnHours?: number;
	powerCycles?: number;
	errors?: string[];
	rawOutput?: string;
}

/**
 * Protected disks that should never be mounted or modified
 */
const PROTECTED_DISKS = ["sdc"]; // Windows partition

/**
 * Disk monitoring utility
 * Monitors disk space, health (SMART), and provides alerts
 */
export class DiskMonitor {
	private static instance: DiskMonitor;
	private logger = Logger.getInstance();

	private constructor() { }

	static getInstance(): DiskMonitor {
		if (!DiskMonitor.instance) {
			DiskMonitor.instance = new DiskMonitor();
		}
		return DiskMonitor.instance;
	}

	/**
	 * Check if a disk is protected (e.g., Windows partition)
	 */
	isProtectedDisk(device: string): boolean {
		return PROTECTED_DISKS.some((protected_disk) =>
			device.includes(protected_disk),
		);
	}

	/**
	 * Get all mounted disk usage information
	 */
	async getAllDiskUsage(): Promise<DiskUsage[]> {
		try {
			// Get disk usage using df command
			const result = await Shell.execute(
				'df -B1 --output=source,target,fstype,size,used,avail,pcent | grep -E "^/dev/"',
			);

			if (result.exitCode !== 0) {
				this.logger.error(`Failed to get disk usage: ${result.stderr}`);
				return [];
			}

			const disks: DiskUsage[] = [];
			const lines = result.stdout.trim().split("\n");

			for (const line of lines) {
				const parts = line.trim().split(/\s+/);
				if (parts.length < 7) continue;

				const [device, mountPoint, filesystem, total, used, free, percent] =
					parts;

				// Validate all required fields exist
				if (!device || !mountPoint || !filesystem || !total || !used || !free || !percent) {
					continue;
				}

				const percentUsed = parseInt(percent.replace("%", ""), 10);
				const totalBytes = parseInt(total, 10);
				const usedBytes = parseInt(used, 10);
				const freeBytes = parseInt(free, 10);

				// Skip if parsing failed
				if (isNaN(percentUsed) || isNaN(totalBytes) || isNaN(usedBytes) || isNaN(freeBytes)) {
					continue;
				}

				// Determine status based on thresholds
				let status: "healthy" | "warning" | "critical" = "healthy";
				if (percentUsed >= 95 || freeBytes < 100 * 1024 * 1024 * 1024) {
					status = "critical";
				} else if (percentUsed >= 90 || freeBytes < 200 * 1024 * 1024 * 1024) {
					status = "warning";
				}

				disks.push({
					device,
					mountPoint,
					filesystem,
					totalBytes,
					usedBytes,
					freeBytes,
					percentUsed,
					totalGB: Math.round((totalBytes / 1024 / 1024 / 1024) * 10) / 10,
					usedGB: Math.round((usedBytes / 1024 / 1024 / 1024) * 10) / 10,
					freeGB: Math.round((freeBytes / 1024 / 1024 / 1024) * 10) / 10,
					status,
				});
			}

			return disks;
		} catch (error) {
			this.logger.error(`Error getting disk usage: ${error}`);
			return [];
		}
	}

	/**
	 * Check for low disk space warnings
	 */
	async checkLowSpace(): Promise<DiskWarning[]> {
		const disks = await this.getAllDiskUsage();
		const warnings: DiskWarning[] = [];

		for (const disk of disks) {
			if (disk.status === "critical") {
				warnings.push({
					device: disk.device,
					mountPoint: disk.mountPoint,
					level: "critical",
					message: `CRITICAL: ${disk.mountPoint} has only ${disk.freeGB}GB free (${disk.percentUsed}% used)`,
					freeGB: disk.freeGB,
					percentUsed: disk.percentUsed,
				});
			} else if (disk.status === "warning") {
				warnings.push({
					device: disk.device,
					mountPoint: disk.mountPoint,
					level: "warning",
					message: `WARNING: ${disk.mountPoint} has ${disk.freeGB}GB free (${disk.percentUsed}% used)`,
					freeGB: disk.freeGB,
					percentUsed: disk.percentUsed,
				});
			}
		}

		return warnings;
	}

	/**
	 * Get SMART health status for a disk
	 * Requires smartmontools (smartctl)
	 */
	async getSmartStatus(device: string): Promise<SmartStatus | null> {
		try {
			// Check if smartctl is available
			const checkResult = await Shell.execute("which smartctl");
			if (checkResult.exitCode !== 0) {
				this.logger.warn("smartctl not found - install smartmontools package");
				return null;
			}

			// Get SMART health
			const result = await Shell.execute(`sudo smartctl -H ${device}`, {});

			const passed = result.stdout.includes("PASSED");
			const status: SmartStatus = {
				device,
				passed,
				rawOutput: result.stdout,
			};

			// Get detailed SMART data
			const detailResult = await Shell.execute(`sudo smartctl -a ${device}`, {});
			if (detailResult.exitCode === 0) {
				// Extract temperature
				const tempMatch = detailResult.stdout.match(
					/Temperature.*?(\d+)\s*Celsius/i,
				);
				if (tempMatch && tempMatch[1]) {
					status.temperature = parseInt(tempMatch[1], 10);
				}

				// Extract power on hours
				const hoursMatch = detailResult.stdout.match(
					/Power_On_Hours.*?(\d+)/i,
				);
				if (hoursMatch && hoursMatch[1]) {
					status.powerOnHours = parseInt(hoursMatch[1], 10);
				}

				// Extract power cycles
				const cyclesMatch = detailResult.stdout.match(
					/Power_Cycle_Count.*?(\d+)/i,
				);
				if (cyclesMatch && cyclesMatch[1]) {
					status.powerCycles = parseInt(cyclesMatch[1], 10);
				}

				// Check for errors
				const errors: string[] = [];
				if (detailResult.stdout.includes("FAILING_NOW")) {
					errors.push("Disk has attributes FAILING NOW");
				}
				if (detailResult.stdout.match(/Reallocated_Sector_Ct.*?\d+/)) {
					const match = detailResult.stdout.match(
						/Reallocated_Sector_Ct\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+(\d+)/,
					);
					if (match && match[1] && parseInt(match[1], 10) > 0) {
						errors.push(`Reallocated sectors: ${match[1]}`);
					}
				}
				if (errors.length > 0) {
					status.errors = errors;
				}
			}

			return status;
		} catch (error) {
			this.logger.error(`Error getting SMART status for ${device}: ${error}`);
			return null;
		}
	}

	/**
	 * Get SMART status for all physical disks
	 */
	async getAllSmartStatus(): Promise<SmartStatus[]> {
		try {
			// Get list of physical disks
			const result = await Shell.execute(
				'lsblk -d -n -o NAME,TYPE | grep disk | awk \'{print "/dev/"$1}\'',
				{},
			);

			if (result.exitCode !== 0) {
				this.logger.error(`Failed to list disks: ${result.stderr}`);
				return [];
			}

			const disks = result.stdout
				.trim()
				.split("\n")
				.filter((d) => d.trim());
			const statuses: SmartStatus[] = [];

			for (const disk of disks) {
				// Skip protected disks
				if (this.isProtectedDisk(disk)) {
					this.logger.info(`Skipping protected disk: ${disk}`);
					continue;
				}

				const status = await this.getSmartStatus(disk);
				if (status) {
					statuses.push(status);
				}
			}

			return statuses;
		} catch (error) {
			this.logger.error(`Error getting SMART status for all disks: ${error}`);
			return [];
		}
	}

	/**
	 * Format disk usage for display
	 */
	formatDiskUsage(disk: DiskUsage): string {
		const statusIcon =
			disk.status === "critical" ? "🔴" : disk.status === "warning" ? "🟡" : "🟢";
		return `${statusIcon} ${disk.mountPoint} (${disk.device}): ${disk.usedGB}GB / ${disk.totalGB}GB (${disk.percentUsed}%) - ${disk.freeGB}GB free`;
	}

	/**
	 * Get a summary of all disk usage
	 */
	async getDiskSummary(): Promise<string> {
		const disks = await this.getAllDiskUsage();
		const warnings = await this.checkLowSpace();

		let summary = "=== Disk Usage Summary ===\n\n";

		// Add warnings first
		if (warnings.length > 0) {
			summary += "⚠️  WARNINGS:\n";
			for (const warning of warnings) {
				summary += `  ${warning.level === "critical" ? "🔴" : "🟡"} ${warning.message}\n`;
			}
			summary += "\n";
		}

		// Add all disks
		summary += "All Disks:\n";
		for (const disk of disks) {
			summary += `  ${this.formatDiskUsage(disk)}\n`;
		}

		return summary;
	}
}

File: src/features/system-health/DiskMonitor.go
/**
 * Disk monitoring utility
 * Monitors disk space, health (SMART), and provides alerts
 */

package systemhealth

import (
	"context"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/ln64-git/daemira/src/utility"
)

// DiskUsage represents disk usage information
type DiskUsage struct {
	Device      string
	MountPoint  string
	Filesystem  string
	TotalBytes  int64
	UsedBytes   int64
	FreeBytes   int64
	PercentUsed float64
	TotalGB     float64
	UsedGB      float64
	FreeGB      float64
	Status      string // "healthy", "warning", "critical"
}

// DiskWarning represents a disk health warning
type DiskWarning struct {
	Device      string
	MountPoint  string
	Level       string // "warning", "critical"
	Message     string
	FreeGB      float64
	PercentUsed float64
}

// SmartStatus represents SMART health status
type SmartStatus struct {
	Device       string
	Passed       bool
	Temperature  *int
	PowerOnHours *int
	PowerCycles  *int
	Errors       []string
	RawOutput    string
}

// Protected disks that should never be mounted or modified
var protectedDisks = []string{"sdc"} // Windows partition

// DiskMonitor monitors disk space, health (SMART), and provides alerts
type DiskMonitor struct {
	logger *utility.Logger
	shell  *utility.Shell
	mu     sync.RWMutex
}

var (
	diskMonitorInstance *DiskMonitor
	diskMonitorOnce     sync.Once
)

// GetDiskMonitor returns the singleton DiskMonitor instance
func GetDiskMonitor() *DiskMonitor {
	diskMonitorOnce.Do(func() {
		diskMonitorInstance = &DiskMonitor{
			logger: utility.GetLogger(),
			shell:  utility.NewShell(utility.GetLogger()),
		}
	})
	return diskMonitorInstance
}

// IsProtectedDisk checks if a disk is protected (e.g., Windows partition)
func (dm *DiskMonitor) IsProtectedDisk(device string) bool {
	for _, protected := range protectedDisks {
		if strings.Contains(device, protected) {
			return true
		}
	}
	return false
}

// GetAllDiskUsage gets all mounted disk usage information
func (dm *DiskMonitor) GetAllDiskUsage(ctx context.Context) ([]DiskUsage, error) {
	result, err := dm.shell.Execute(ctx,
		`df -B1 --output=source,target,fstype,size,used,avail,pcent | grep -E "^/dev/"`,
		&utility.ExecOptions{
			Timeout: 10 * time.Second,
		})

	if err != nil || result.ExitCode != 0 {
		dm.logger.Error("Failed to get disk usage: %v", err)
		return []DiskUsage{}, err
	}

	var disks []DiskUsage
	lines := strings.Split(strings.TrimSpace(result.Stdout), "\n")

	for _, line := range lines {
		parts := strings.Fields(strings.TrimSpace(line))
		if len(parts) < 7 {
			continue
		}

		device := parts[0]
		mountPoint := parts[1]
		filesystem := parts[2]
		totalStr := parts[3]
		usedStr := parts[4]
		freeStr := parts[5]
		percentStr := parts[6]

		// Validate all required fields exist
		if device == "" || mountPoint == "" || filesystem == "" ||
			totalStr == "" || usedStr == "" || freeStr == "" || percentStr == "" {
			continue
		}

		percentUsed, err := strconv.ParseFloat(strings.TrimSuffix(percentStr, "%"), 64)
		if err != nil {
			continue
		}

		totalBytes, err := strconv.ParseInt(totalStr, 10, 64)
		if err != nil {
			continue
		}

		usedBytes, err := strconv.ParseInt(usedStr, 10, 64)
		if err != nil {
			continue
		}

		freeBytes, err := strconv.ParseInt(freeStr, 10, 64)
		if err != nil {
			continue
		}

		// Determine status based on thresholds
		status := "healthy"
		if percentUsed >= 95 || freeBytes < 100*1024*1024*1024 {
			status = "critical"
		} else if percentUsed >= 90 || freeBytes < 200*1024*1024*1024 {
			status = "warning"
		}

		disks = append(disks, DiskUsage{
			Device:      device,
			MountPoint:  mountPoint,
			Filesystem:  filesystem,
			TotalBytes:  totalBytes,
			UsedBytes:   usedBytes,
			FreeBytes:   freeBytes,
			PercentUsed: percentUsed,
			TotalGB:     float64(totalBytes) / 1024 / 1024 / 1024,
			UsedGB:      float64(usedBytes) / 1024 / 1024 / 1024,
			FreeGB:      float64(freeBytes) / 1024 / 1024 / 1024,
			Status:      status,
		})
	}

	return disks, nil
}

// CheckLowSpace checks for low disk space warnings
func (dm *DiskMonitor) CheckLowSpace(ctx context.Context) ([]DiskWarning, error) {
	disks, err := dm.GetAllDiskUsage(ctx)
	if err != nil {
		return nil, err
	}

	var warnings []DiskWarning
	for _, disk := range disks {
		if disk.Status == "critical" {
			warnings = append(warnings, DiskWarning{
				Device:      disk.Device,
				MountPoint:  disk.MountPoint,
				Level:       "critical",
				Message:     fmt.Sprintf("CRITICAL: %s has only %.1fGB free (%.1f%% used)", disk.MountPoint, disk.FreeGB, disk.PercentUsed),
				FreeGB:      disk.FreeGB,
				PercentUsed: disk.PercentUsed,
			})
		} else if disk.Status == "warning" {
			warnings = append(warnings, DiskWarning{
				Device:      disk.Device,
				MountPoint:  disk.MountPoint,
				Level:       "warning",
				Message:     fmt.Sprintf("WARNING: %s has %.1fGB free (%.1f%% used)", disk.MountPoint, disk.FreeGB, disk.PercentUsed),
				FreeGB:      disk.FreeGB,
				PercentUsed: disk.PercentUsed,
			})
		}
	}

	return warnings, nil
}

// GetSmartStatus gets SMART health status for a disk
// Requires smartmontools (smartctl)
func (dm *DiskMonitor) GetSmartStatus(ctx context.Context, device string) (*SmartStatus, error) {
	// Check if smartctl is available
	checkResult, err := dm.shell.Execute(ctx, "which smartctl", &utility.ExecOptions{
		Timeout: 2 * time.Second,
	})
	if err != nil || checkResult.ExitCode != 0 {
		dm.logger.Warn("smartctl not found - install smartmontools package")
		return nil, fmt.Errorf("smartctl not available")
	}

	// Get SMART health
	result, err := dm.shell.Execute(ctx, fmt.Sprintf("sudo smartctl -H %s", device), &utility.ExecOptions{
		Timeout: 30 * time.Second,
	})
	if err != nil {
		return nil, err
	}

	passed := strings.Contains(result.Stdout, "PASSED")
	status := &SmartStatus{
		Device:    device,
		Passed:    passed,
		RawOutput: result.Stdout,
	}

	// Get detailed SMART data
	detailResult, err := dm.shell.Execute(ctx, fmt.Sprintf("sudo smartctl -a %s", device), &utility.ExecOptions{
		Timeout: 30 * time.Second,
	})
	if err == nil && detailResult.ExitCode == 0 {
		// Extract temperature
		tempRegex := regexp.MustCompile(`Temperature.*?(\d+)\s*Celsius`)
		if matches := tempRegex.FindStringSubmatch(detailResult.Stdout); len(matches) > 1 {
			if temp, err := strconv.Atoi(matches[1]); err == nil {
				status.Temperature = &temp
			}
		}

		// Extract power on hours
		hoursRegex := regexp.MustCompile(`Power_On_Hours.*?(\d+)`)
		if matches := hoursRegex.FindStringSubmatch(detailResult.Stdout); len(matches) > 1 {
			if hours, err := strconv.Atoi(matches[1]); err == nil {
				status.PowerOnHours = &hours
			}
		}

		// Extract power cycles
		cyclesRegex := regexp.MustCompile(`Power_Cycle_Count.*?(\d+)`)
		if matches := cyclesRegex.FindStringSubmatch(detailResult.Stdout); len(matches) > 1 {
			if cycles, err := strconv.Atoi(matches[1]); err == nil {
				status.PowerCycles = &cycles
			}
		}

		// Check for errors
		var errors []string
		if strings.Contains(detailResult.Stdout, "FAILING_NOW") {
			errors = append(errors, "Disk has attributes FAILING NOW")
		}

		reallocRegex := regexp.MustCompile(`Reallocated_Sector_Ct\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+(\d+)`)
		if matches := reallocRegex.FindStringSubmatch(detailResult.Stdout); len(matches) > 1 {
			if count, err := strconv.Atoi(matches[1]); err == nil && count > 0 {
				errors = append(errors, fmt.Sprintf("Reallocated sectors: %d", count))
			}
		}

		if len(errors) > 0 {
			status.Errors = errors
		}
	}

	return status, nil
}

// GetAllSmartStatus gets SMART status for all physical disks
func (dm *DiskMonitor) GetAllSmartStatus(ctx context.Context) ([]SmartStatus, error) {
	result, err := dm.shell.Execute(ctx,
		`lsblk -d -n -o NAME,TYPE | grep disk | awk '{print "/dev/"$1}'`,
		&utility.ExecOptions{
			Timeout: 10 * time.Second,
		})

	if err != nil || result.ExitCode != 0 {
		dm.logger.Error("Failed to list disks: %v", err)
		return []SmartStatus{}, err
	}

	disks := strings.Fields(strings.TrimSpace(result.Stdout))
	var statuses []SmartStatus

	for _, disk := range disks {
		// Skip protected disks
		if dm.IsProtectedDisk(disk) {
			dm.logger.Info("Skipping protected disk: %s", disk)
			continue
		}

		status, err := dm.GetSmartStatus(ctx, disk)
		if err != nil {
			continue
		}
		if status != nil {
			statuses = append(statuses, *status)
		}
	}

	return statuses, nil
}

// FormatDiskUsage formats disk usage for display
func (dm *DiskMonitor) FormatDiskUsage(disk DiskUsage) string {
	var statusIcon string
	switch disk.Status {
	case "critical":
		statusIcon = "🔴"
	case "warning":
		statusIcon = "🟡"
	default:
		statusIcon = "🟢"
	}
	return fmt.Sprintf("%s %s (%s): %.1fGB / %.1fGB (%.1f%%) - %.1fGB free",
		statusIcon, disk.MountPoint, disk.Device, disk.UsedGB, disk.TotalGB, disk.PercentUsed, disk.FreeGB)
}

// GetDiskSummary gets a summary of all disk usage
func (dm *DiskMonitor) GetDiskSummary(ctx context.Context) (string, error) {
	disks, err := dm.GetAllDiskUsage(ctx)
	if err != nil {
		return "", err
	}

	warnings, err := dm.CheckLowSpace(ctx)
	if err != nil {
		return "", err
	}

	summary := "=== Disk Usage Summary ===\n\n"

	// Add warnings first
	if len(warnings) > 0 {
		summary += "⚠️  WARNINGS:\n"
		for _, warning := range warnings {
			icon := "🟡"
			if warning.Level == "critical" {
				icon = "🔴"
			}
			summary += fmt.Sprintf("  %s %s\n", icon, warning.Message)
		}
		summary += "\n"
	}

	// Add all disks
	summary += "All Disks:\n"
	for _, disk := range disks {
		summary += fmt.Sprintf("  %s\n", dm.FormatDiskUsage(disk))
	}

	return summary, nil
}
